src/async_fuse/memfs/kv_engine/etcd_impl.rs

```diff
@@ -1,19 +1,191 @@
+        self.inner.lease_id
+    }
+
+    /// Get the ttl
+    pub fn ttl(&self) -> i64 {
+        self.inner.ttl
+    }
+
+    /// Get the is_closed flag
+    pub fn is_closed(&self) -> bool {
+        self.inner
+            .is_closed
+            .load(std::sync::atomic::Ordering::Relaxed)
+    }
+}
+
+impl Drop for Session {
+    fn drop(&mut self) {
+        // Try to close the session
+        let _ = self.close_tx.try_send(());
+        // Set the is_closed flag
+        self.inner
+            .is_closed
+            .store(true, std::sync::atomic::Ordering::Relaxed);
+    }
+}
+
+impl SessionInner {
+    /// Keep alive the lease
+    pub async fn keep_alive(&self, mut close_rx: mpsc::Receiver<()>) -> DatenLordResult<()> {
+        let mut interval = tokio::time::interval(Duration::from_secs(self.ttl as u64 / 3));
+
+        loop {
+            if self.is_closed.load(std::sync::atomic::Ordering::Relaxed) {
+                error!("lease keep alive stream closed by is_closed flag");
+                return Ok(());
+            }
+
+            // Try to clone a client, if the keeper failed, try to reconnect
+            let mut client = self.client.clone();
+            let lease_id = self.lease_id;
+            let (mut keeper, mut lease_keep_alive_stream) =
+                match client.lease_keep_alive(lease_id).await {
+                    Ok((keeper, lease_keep_alive_stream)) => (keeper, lease_keep_alive_stream),
+                    Err(e) => {
+                        error!("failed to keep alive lease, error={e:?}");
+
+                        // Retry to connect
+                        tokio::time::sleep(Duration::from_secs(1)).await;
+                        continue;
+                    }
+                };
+
+            // Set the is_closed flag
+            self.is_closed
+                .store(false, std::sync::atomic::Ordering::Relaxed);
+
+            // Start to keep alive the lease
+            loop {
+                if self.is_closed.load(std::sync::atomic::Ordering::Relaxed) {
+                    error!("lease keep alive stream closed by is_closed flag");
+                    return Ok(());
+                }
+
+                tokio::select! {
+                    _ = interval.tick() => {
+                        // Try to send a keep alive request
+                        match keeper.keep_alive().await {
+                            Ok(_) => {
+                                info!("keep alive lease, lease_id={lease_id}");
+                            }
+                            Err(e) => {
+                                error!("failed to keep alive lease, error={e:?}");
+                                break;
+                            }
+                        }
+
+                        // Try to parse
+                        match lease_keep_alive_stream.message().await {
+                            Ok(Some(val)) => {
+                                if val.ttl() == 0 {
```

Comment:
`if val.expired()` is a better name?
And the name of 'val` is lack of information, better renmae it.

Reply:
Your are right, a `expired` can be used to check stream is simple to read.

Reply:
`ttl` is a library function defined in `/home/lvbo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/etcd-client-0.11.1/src/rpc/lease.rs`.